pkg/destroy/azure/azure.go

```diff
@@ -226,9 +247,90 @@ func (o *ClusterUninstaller) Run() (*types.ClusterQuota, error) {
 		o.Logger.Debug(err)
 	}
 
+	deadline, _ = waitCtx.Deadline()
+	diff = time.Until(deadline)
+	if diff > 0 {
+		waitCtx, cancel = context.WithTimeout(context.Background(), diff)
+		defer cancel()
+	}
+
+	if err := removeSharedTags(
+		waitCtx, o.resourceGraphClient, o.tagsClient, o.InfraID, o.SubscriptionID,
+	); err != nil {
+		errs = append(errs, fmt.Errorf("failed to remove shared tags: %w", err))
+		o.Logger.Debug(err)
+	}
+
 	return nil, utilerrors.NewAggregate(errs)
 }
 
+func removeSharedTags(
+	ctx context.Context,
+	graphClient *armresourcegraph.Client,
+	tagsClient *armresources.TagsClient,
+	infraID, subscriptionID string,
+) error {
+	tagKey := fmt.Sprintf("kubernetes.io_cluster.%s", infraID)
+	query := fmt.Sprintf(
+		"resources | where type == 'microsoft.network/virtualnetworks' and tags.['%s'] == 'shared' | project name, resourceGroup, tags",
+		tagKey,
+	)
+	results, err := graphClient.Resources(ctx,
+		armresourcegraph.QueryRequest{
+			Query: &query,
+			Subscriptions: []*string{
+				&subscriptionID,
+			},
+		},
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to query vnet: %w", err)
+	}
+
+	// marshal and unmarshal results.Data to avoid casting every type along the way.
+	// the sdk forcing us to do this is a shortcoming on its part which has been
+	// reported but so far no solution have been provided.
+	// see https://github.com/Azure/azure-sdk-for-go/issues/21004 for details.
+	b, err := json.Marshal(results.Data)
+	if err != nil {
+		return fmt.Errorf("failed to json marshall results data: %s", results.Data)
+	}
+	vnetsResult := []struct {
+		Name          string
+		ResourceGroup string
+		Tags          map[string]string
+	}{}
+	if err := json.Unmarshal(b, &vnetsResult); err != nil {
+		return fmt.Errorf("failed to json marshall results data: %s", results.Data)
+	}
+
+	for _, vnet := range vnetsResult {
+		scope := fmt.Sprintf(
+			"subscriptions/%s/resourcegroups/%s/providers/Microsoft.Network/virtualNetworks/%s",
+			subscriptionID,
+			vnet.ResourceGroup,
+			vnet.Name,
+		)
+
+		deleteOp := armresources.TagsPatchOperationDelete
+		tagsParam := armresources.TagsPatchResource{
+			Operation: &deleteOp,
+			Properties: &armresources.Tags{
+				Tags: map[string]*string{
+					tagKey: to.StringPtr("shared"),
+				},
+			},
+		}
+
+		if _, err := tagsClient.UpdateAtScope(ctx, scope, tagsParam, nil); err != nil {
+			return fmt.Errorf("failed to remove shared tag from vnet: %w", err)
+		}
+	}
+
+	return nil
```

Comment:
What about:
```suggestion
	query := fmt.Sprintf(
		"resources | where tags.['%s'] == 'shared' | project id, name, type",
		tagKey,
	)
	results, err := graphClient.Resources(ctx,
		armresourcegraph.QueryRequest{
			Query: azcoreto.Ptr(query),
			Subscriptions: []*string{
				azcoreto.Ptr(subscriptionID),
			},
			Options: &armresourcegraph.QueryRequestOptions{
				ResultFormat: azcoreto.Ptr(armresourcegraph.ResultFormatObjectArray),
			},
		},
		nil,
	)
	if err != nil {
		return fmt.Errorf("failed to query resources with shared tag: %w", err)
	}

	deleteOp := armresources.TagsPatchOperationDelete
	tagsParam := armresources.TagsPatchResource{
		Operation: &deleteOp,
		Properties: &armresources.Tags{
			Tags: map[string]*string{
				tagKey: to.StringPtr("shared"),
			},
		},
	}

	var errs []error
	if m, ok := results.Data.([]any); ok {
		for _, r := range m {
			items := r.(map[string]any)
			resourceName := items["name"].(string)
			resourceType := items["type"].(string)
			resourceID := items["id"].(string)
			logger := logger.WithFields(logrus.Fields{
				"resource": resourceName,
				"type":     resourceType,
			})
			logger.Debugf("removing shared tag from resource")
			if _, err := tagsClient.UpdateAtScope(ctx, resourceID, tagsParam, nil); err != nil {
				errs = append(errs, fmt.Errorf("failed to remove shared tag from %s: %w", items["name"], err))
			}
			logger.Infoln("removed shared tag")
		}
	}

	return utilerrors.NewAggregate(errs)
```

Reply:
Oh so that's what you mean by "more generic"! I get it now :) will do.

Reply:
yeah. Sorry for not being more specific before, but I was not very familiar with graph queries. I'm still getting the grasp of the new Azure sdk.

Reply:
done: https://github.com/openshift/installer/pull/7611/commits/489c0ec7c9cb75fe557fe180c4fee336460310fb

Reply:
no worries, and same here.